exec/java-exec/src/main/java/org/apache/drill/exec/util/filereader/DirectBufInputStream.java

```diff
@@ -0,0 +1,166 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.util.filereader;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.DrillBuf;
+import org.apache.drill.exec.memory.BufferAllocator;
+import org.apache.hadoop.fs.ByteBufferReadable;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.parquet.hadoop.util.CompatibilityUtil;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+
+public class DirectBufInputStream extends FilterInputStream {
+
+  private static final org.slf4j.Logger logger =
+      org.slf4j.LoggerFactory.getLogger(DirectBufInputStream.class);
+
+  protected boolean enableHints = true;
+  protected String streamId; // a name for logging purposes only
+  protected BufferAllocator allocator;
+  /**
+   * The length of the data we expect to read. The caller may, in fact,
+   * ask for more or less bytes. However this is useful for providing hints where
+   * the underlying InputStream supports hints (e.g. fadvise)
+   */
+  protected final long totalByteSize;
+
+  /**
+   * The offset in the underlying stream to start reading from
+   */
+  protected final long startOffset;
+
+  public DirectBufInputStream(InputStream in, BufferAllocator allocator, String id, long startOffset,
+      long totalByteSize, boolean enableHints) {
+    super(in);
+    Preconditions.checkArgument(startOffset >= 0);
+    Preconditions.checkArgument(totalByteSize >= 0);
+    this.streamId = id;
+    this.allocator = allocator;
+    this.startOffset = startOffset;
+    this.totalByteSize = totalByteSize;
+    this.enableHints = enableHints;
+  }
+
+  public void init() throws IOException, UnsupportedOperationException {
+    checkStreamSupportsByteBuffer();
+    if (enableHints) {
+      fadviseIfAvailable(getInputStream(), this.startOffset, this.totalByteSize);
+    }
+    getInputStream().seek(this.startOffset);
+    return;
+  }
+
+  public int read() throws IOException {
+    return getInputStream().read();
+  }
+
+  public synchronized int read(DrillBuf buf, int off, int len) throws IOException {
+    buf.clear();
+    ByteBuffer directBuffer = buf.nioBuffer(0, len);
```

Comment:
Seems to create garbage on every read (to create the ByteBuffer with the proper offset and length.) Any way to do the read without garbage?